pkg/util/lock.go

```diff
@@ -111,81 +114,175 @@ func (l *PgAdvisoryLock) TryLock() (bool, error) {
 	return true, nil
 }
 
-func (l *PgAdvisoryLock) getAdvisoryLock() (bool, error) {
-	var err error
-	if l.conn == nil {
-		l.conn, err = getConn(l.connStr, 0, 10)
+// Resign releases the leader status of this instance.
+func (l *PgLeaderLock) Resign() error {
+	return l.Release()
+}
+
+// Locked returns if the instance was able to obtain the locks.
+func (l *PgLeaderLock) Locked() bool {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+	return l.obtained
+}
+
+// Release releases the already obtained locks.
+func (l *PgLeaderLock) Release() error {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	defer l.connCleanUp()
+	if !l.obtained {
+		return fmt.Errorf("can't release while not holding the lock")
 	}
+
+	defer func() { l.obtained = false }()
+
+	unlocked, err := l.unlock()
 	if err != nil {
-		return false, err
+		return err
 	}
-	defer func() {
-		if err != nil {
-			l.connCleanUp()
-		}
-	}()
-	rows, err := l.conn.Query(context.Background(), "SELECT pg_try_advisory_lock($1)", l.groupLockID)
+	if !unlocked {
+		log.Debug("msg", fmt.Sprintf("false release for group id %d", l.groupLockID))
+	}
+
+	return nil
+}
+
+type PgAdvisoryLock struct {
+	conn         *pgx.Conn
+	connStr      string
+	groupLockID  int64
+	afterConnect AfterConnectFunc
+
+	mutex sync.RWMutex
+}
+
+type AdvisoryLock interface {
+	GetAdvisoryLock() (bool, error)
+	GetSharedAdvisoryLock() (bool, error)
+	Unlock() (bool, error)
+	UnlockShared() (bool, error)
+	Close()
+}
+
+// PgAdvisoryLock is a AdvisoryLock
+var _ AdvisoryLock = (*PgAdvisoryLock)(nil)
+
+// NewPgAdvisoryLock creates a new instance with specified lock ID, connection pool and lock timeout.
+func NewPgAdvisoryLock(groupLockID int64, connStr string) (*PgAdvisoryLock, error) {
+	lock := &PgAdvisoryLock{
+		connStr:     connStr,
+		groupLockID: groupLockID,
+	}
+	return lock, nil
+}
+
+func (l *PgAdvisoryLock) getConn(connStr string, cur, maxRetries int) (*pgx.Conn, error) {
+	if maxRetries == cur {
+		return nil, fmt.Errorf("max attempts reached. giving up on getting a db connection")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), waitForConnectionTimeout)
+	defer cancel()
+
+	cfg, err := pgx.ParseConfig(connStr)
 	if err != nil {
-		return false, err
+		return nil, fmt.Errorf("error parsing config connection: %w", err)
 	}
-	defer rows.Close()
-	if !rows.Next() {
-		err = rows.Err()
+
```

Comment:
question: why don't we use cfg.AfterConnect here instead?

Reply:
it's got a weird function type intended for much more low-level use than we want (it exposes the underlying `PgConn`)